### DISCUSSION TITLE
[Task]: How to change threshold in padin config.yaml ??

### DISCUSSION THREAD

**Discussion by RDADADA** [open-edge-platform/anomalib#1434] [Q&A] [answered]
### What is the motivation for this task?

1. The inferred defects appear clearly on the heat map, but the result is not circled.
    I think the threshold is not adjusted properly, so I want to know how to adjust the threshold.

2. this is my custom yaml

dataset:
  name: ICtray_seg
  format: folder
  root: ./datasets/ICtray_seg
  normal_dir: train/good
  abnormal_dir: test/post_fake
  normal_test_dir: test/good
  mask_dir: ground_truth/post_fake
  task: segmentation
  extensions: null
  train_batch_size: 4
  test_batch_size: 4
  num_workers: 8
  image_size:
  - 256
  - 256
  center_crop: null
  normalization: imagenet
  transform_config:
    train: null
    eval: null
  test_split_mode: from_dir
  test_split_ratio: 0.2
  val_split_mode: same_as_test
  val_split_ratio: 0.5
  tiling:
    apply: false
    tile_size: null
    stride: null
    remove_border_count: 0
    use_random_tiling: false
    random_tile_count: 16
  eval_batch_size: 4
model:
  name: padim
  backbone: resnet18
  pre_trained: true
  layers:
    - layer1
    - layer2
    - layer3
  normalization_method: min_max # options: [none, min_max, cdf]
metrics:
  image:
  - F1Score
  - AUROC
  pixel:
  - F1Score
  - AUROC
  threshold:
    method: adaptive
    manual_image: null
    manual_pixel: null
visualization:
  show_images: false
  save_images: true
  log_images: true
  image_save_path: null
  mode: full
project:
  seed: 0
  path: results/padim/ICtray_seg/run
  unique_dir: false
logging:
  logger: []
  log_graph: false
optimization:
  export_mode: null
trainer:
  enable_checkpointing: true
  default_root_dir: results/padim/ICtray_seg/run
  gradient_clip_val: 0
  gradient_clip_algorithm: norm
  num_nodes: 1
  devices: 1
  enable_progress_bar: true
  overfit_batches: 0.0
  track_grad_norm: -1
  check_val_every_n_epoch: 1
  fast_dev_run: false
  accumulate_grad_batches: 1
  max_epochs: 1
  min_epochs: null
  max_steps: -1
  min_steps: null
  max_time: null
  limit_train_batches: 1.0
  limit_val_batches: 1.0
  limit_test_batches: 1.0
  limit_predict_batches: 1.0
  val_check_interval: 1.0
  log_every_n_steps: 50
  accelerator: auto
  strategy: null
  sync_batchnorm: false
  precision: 32
  enable_model_summary: true
  num_sanity_val_steps: 0
  profiler: null
  benchmark: false
  deterministic: false
  reload_dataloaders_every_n_epochs: 0
  auto_lr_find: false
  replace_sampler_ddp: true
  detect_anomaly: false
  auto_scale_batch_size: false
  plugins: null
  move_metrics_to_cpu: false
  multiple_trainloader_mode: max_size_cycle

### Describe the solution you'd like

I want to know how to change threshold by my mamual value, if you can give me a example, i wll very appreciate that.

  threshold:
    method: manual
    manual_image: null (is a path of image or what?)
    manual_pixel: null (is a pixel value or position in the manual_image ? )

**Comment by blaz-r** [comment, marked as answer]
Hello.
You adjust it like this:
```yaml
metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    method: manual          # set to manual so it's not calculated automatically
    manual_image: 0.42      # provide threshold value
    manual_pixel: 0.42      # provide threshold value
```

So by setting method to manual and providing values.

**Reply by RDADADA** [reply]
Hi,
thanks for your explanation, this is an another question
I used the threshold like this:

method: manual         
    manual_image: 0.42      
    manual_pixel: 0.42      

I try 0.01 to 0.9 for  both of  manual_image and manual_pixel 
the predicted heat map got my defect, but predicted mask is none and the result not marked defects.

The predicted result can work when threshold method is adaptive
Is there any tool any get the value of manual_image and manual_pixel when  threshold method is adaptive?

**Reply by blaz-r** [reply]
If you don't use normalization, then the threshold can also be greater than 1. In your case it seems like some other issue which I'm not sure about.

You can modify the code to print those thresholds or you can use a debugger to view it.

**Reply by RDADADA** [reply]
Hi
can I ask where is the code make heat map to predicted mask in lightning_inference.py ?
I want to figure out why heat map can find the defect but predicted mask is none.
appreciate that a lot.

**Reply by blaz-r** [reply]
It's here in line 95:
https://github.com/openvinotoolkit/anomalib/blob/66532fc1625020ce7ff6f3fc1565663aa540e834/src/anomalib/models/components/base/anomaly_module.py#L74-L106

**Reply by RDADADA** [reply]
Thank you for this imformation, it hleps me a lot
